### PR TITLE
Add the ability for extensions to exclude GraalVM config from jars

### LINKED ISSUE #17482: Pass --exclude-config programatically
`--exclude-config` was introduced in https://github.com/oracle/graal/pull/3179
It would be nice to be able to pass it programatically from a Quarkus extension

### PR DESCRIPTION
**Pull request by geoand**
Resolves: #17482

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 1]
```diff
@@ -694,12 +705,18 @@ public NativeImageInvokerInfo build() {
...
-                                .map(p -> p.getSecurityProvider())
+                                .map(NativeImageSecurityProviderBuildItem::getSecurityProvider)
                                 .collect(Collectors.joining(","));
                         nativeImageArgs.add("-H:AdditionalSecurityProviders=" + additionalSecurityProviders);
                     }
+
+                    // Add --exclude-config options
+                    for (ExcludeConfigBuildItem excludeConfig : excludeConfigs) {
+                        nativeImageArgs.add("--exclude-config");
+                        nativeImageArgs.add(excludeConfig.getValue());
```
Comment (jaikiran) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:718]: Hello @geoand, looking at the PR in graalvm that introduced this feature, I think the original comment/description (the top most one) in that PR is no longer accurate. From what I can see, that was the initial proposal which then got modified and from the looks of it, the way to configure this is now a "2 arg value" for that `--exclude-config`. Something like:

`--exclude-config <jar-file-name-or-regex> <resource-name-or-regex-within-that-jar>`

and the 2 args are mandatory. So I think this new `ExcludeConfigBuildItem` would need to be enhanced to allow that.

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:718]: Thanks @jaikiran! You are almost certainly right. 
I looked at the code a little and even tried to debug it, but for some reason I could not get the breakpoint in the code of the PR to hit so I was in the dark :)

I will update the PR based on your observation

**Comment by jaikiran** [comment]
At a high level, keeping aside the 2 arg value comment that I added, I think we should perhaps make this `ExcludeConfigBuildItem` a bit more usable for extension developers, if possible. What I mean is, the value they provide to this build item is expected to be a jar file name (or regex). Expecting them to hardcode/pass these jar file names or regex I think could be brittle (because they then have to go check how the dependency jar name looks like, does it have a version included in its name etc...). Perhaps we should allow them to pass a `Class` instance to this build item and we internally then find the `java.security.CodeSource` of that specific class and then apply the `--exclude-config` against that jar name (which we get dynamically via the `CodeSource`)? For example (a completely random one), if they want to exclude a MySQL driver jar's native configuration resource, then instead of knowing that jar's name, they could just specify the `Class` instance of the MySQL driver (or maybe even the classname?) and we then go find the right jar that needs to be excluded. I think expecting a Class (name) or some such thing in the build item will probably make this much more robust and easy to use than jar names.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java]: Fixed

**Comment by geoand** [comment]
@jaikiran yeah that makes sense, but I would first like @Sanne and @ppalaga to give it a try with the use cases they have in mind. 
If it indeed turns out to be burden to specify strings, we can certainly make it easier using things like you mention

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java]: Thanks a lot!

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package io.quarkus.deployment.builditem.nativeimage;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * A build item that allows extension to configure the native-image compiler to effectively
+ * ignore certain configuration files in specific jars.
+ *
+ * The {@code jarFile} property specifies the name of the jar file or a regular expression that can be used to
```
Comment (Sanne) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java:9]: No need to remove this ability, but I hope we'll never need regexes to match them :)

While we need this safeguard, we should not encourage using it if not as last resort.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 3]
```diff
@@ -694,12 +705,19 @@ public NativeImageInvokerInfo build() {
                     nativeImageArgs.add("-H:-ParseOnce");
 
                     // AdditionalSecurityProviders
-                    if (nativeImageSecurityProviders != null && !nativeImageSecurityProviders.isEmpty()) {
+                    if (!nativeImageSecurityProviders.isEmpty()) {
```
Comment (Sanne) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:708]: Was this meant to be included in this PR? Doesn't seem related

**Review by Sanne — CHANGES_REQUESTED** [review]
Looks good in principle, but there's some changes that seem unrelated.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * For the match to work, the resources need to be part of the matched jar file(s) (see {@code jarFile})
+ *
+ * See https://github.com/oracle/graal/pull/3179 for more details.
+ */
+public final class ExcludeConfigBuildItem extends MultiBuildItem {
+
+    private final String jarFile;
+    private final String resourceName;
+
+    public ExcludeConfigBuildItem(String jarFile, String resourceName) {
```
Comment (jaikiran) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java:23]: For these `String` version of the constructor, perhaps we should have an additional constructor which just takes the jar file name/regex and defaults the resource name to `META-INF/native-image/native-image.properties` (we need to experiment/test to see if that requires a leading `/`, but you get the idea), because most of these libraries which package these configs would be packaging it in this specific resource since that's what's used by the graalvm while looking for these configs[1]

[1] https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/metainf/NativeImageMetaInfWalker.java#L43

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java (hunk 3 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:708]: Yeah, the null check is not needed. I can remove this change if you like

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java]: Removed

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java (hunk 4 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java:25]: Good idea, done

Comment (jaikiran) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java:25]: Thank you. Looks good.

The only thing that I'm a bit uncertain is whether that `META-INF/...` path will need a leading `/` slash for it to work. But that part can only be confirmed by a test case. I don't think it's easy to add that test case and test the native image generation, until some processor/extension starts using this build item. So I think this is good enough for me in its current form

**Review by jaikiran — APPROVED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java (hunk 4 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/nativeimage/ExcludeConfigBuildItem.java:25]: Yeah, exactly.

Looking at the examples from Galder's PR, it shouldn't be needed, but like you said, we can only be sure with a proper test

**Review by ppalaga — APPROVED** [review]
The PR looks good, thanks. I do not think we have anything waiting urgently for this ATM.

**Review by Sanne — APPROVED** [review]
@geoand yes sure. I was hoping to try it out as well, but couldn't get to that yet - I'll do it soon but it's no reason to hold you back.
